h/auth/policy.py

```diff
@@ -261,33 +262,21 @@ def check(username, password, request):
         :returns: additional principals for the auth_client or None
         :rtype: list or None
         """
-        client_id = username
-        client_secret = password
-
         # validate that the credentials in BasicAuth header
         # match an AuthClient record in the db
-        client = util.verify_auth_client(client_id, client_secret, request.db)
-
-        if client is None:
-            return None
-
-        forwarded_userid = AuthClientPolicy._forwarded_userid(request)
-
-        if (
-            forwarded_userid is None
-        ):  # No forwarded user; set principals for basic auth_client
-            return util.principals_for_auth_client(client)
 
-        user_service = request.find_service(name="user")
-        try:
-            user = user_service.fetch(forwarded_userid)
-        except InvalidUserId:  # raised if userid is invalidly formatted
-            return None  # invalid user, so we are failing here
+        client = util.verify_auth_client(
+            client_id=username, client_secret=password, db_session=request.db
+        )
+        user = None
 
-        if user and user.authority == client.authority:
-            return util.principals_for_auth_client_user(user, client)
+        if client and (forwarded_userid := AuthClientPolicy._forwarded_userid(request)):
+            try:
+                user = request.find_service(name="user").fetch(forwarded_userid)
+            except InvalidUserId:  # raised if userid is invalidly formatted
+                return None  # invalid user, so we are failing here
```

Comment:
The refactor here is to build the correct identity object and let the `principals_for_identity` take care of the rest.

In particular the authority matching check has moved to the `principals_for_identity` method.